lib/execution_segment_test.go

```diff
@@ -379,4 +386,107 @@ func TestExecutionSegmentStringSequences(t *testing.T) {
 	}
 }
 
-// TODO: test with randomized things
+// Return a randomly distributed sequence of numSegments amount of
+// execution segments whose length totals 1.
+func genRandomExecutionSegmentSequence(numSegments int) (ExecutionSegmentSequence, error) {
+	const denom int = 1000
+
+	bounds := make(map[int]struct{}, numSegments-1)
+	for i := 0; i < numSegments-1; i++ {
+		b := rand.Intn(denom-1) + 1
+		// Avoid duplicates
+		if _, ok := bounds[b]; ok {
+			numSegments++
+			continue
+		}
+		bounds[b] = struct{}{}
+	}
+
+	nums := make([]int, 0, len(bounds)+2)
+	for k := range bounds {
+		nums = append(nums, k)
+	}
+	nums = append(nums, []int{0, denom}...)
+
+	sort.Ints(nums)
+
+	segments := make([]*ExecutionSegment, 0, len(bounds)+1)
+	denom64 := int64(denom)
+	for i := 0; i < len(nums)-1; i++ {
+		from, to := big.NewRat(int64(nums[i]), denom64), big.NewRat(int64(nums[i+1]), denom64)
+		segment, err := NewExecutionSegment(from, to)
+		if err != nil {
+			return nil, err
+		}
+		segments = append(segments, segment)
+	}
+
+	return NewExecutionSegmentSequence(segments...)
+}
+
+// Ensure that the sum of scaling all execution segments in
+// the same sequence with scaling factor M results in M itself.
+func TestExecutionSegmentScaleConsistency(t *testing.T) {
+	t.Parallel()
+
+	const numTests = 10
+	for i := 0; i < numTests; i++ {
+		scale := rand.Int31n(99) + 2
+		seq, err := genRandomExecutionSegmentSequence(rand.Intn(9) + 2)
+		require.NoError(t, err)
+
+		t.Run(fmt.Sprintf("%d_%s", scale, seq), func(t *testing.T) {
+			var total int64
+			for _, segment := range seq {
+				total += segment.Scale(int64(scale))
+			}
+			assert.Equal(t, int64(scale), total)
+		})
+	}
+}
+
+func TestExecutionSegmentScaleNoWobble(t *testing.T) {
+	t.Skip("disabled until the implementation is fixed")
+	t.Parallel()
+
+	requireSegmentScaleGreater := func(t *testing.T, es *ExecutionSegment) {
+		var i, lastResult int64
+		for i = 1; i < 1000; i++ {
+			result := es.Scale(i)
+			require.GreaterOrEqual(t, result, lastResult)
+			lastResult = result
+		}
+	}
+
+	// Baseline full segment test
+	t.Run("0:1", func(t *testing.T) {
+		es, err := NewExecutionSegment(zeroRat, oneRat)
+		require.NoError(t, err)
+		requireSegmentScaleGreater(t, es)
+	})
+
+	// Random segments
+	const numTests = 10
+	for i := 0; i < numTests; i++ {
+		seq, err := genRandomExecutionSegmentSequence(rand.Intn(3) + 2)
+		require.NoError(t, err)
+
+		es := seq[rand.Intn(len(seq))]
+		t.Run(es.String(), func(t *testing.T) {
+			requireSegmentScaleGreater(t, es)
+		})
+	}
+}
+
+func BenchmarkExecutionSegmentScale(b *testing.B) {
+	es, err := NewExecutionSegmentFromString("73/716:431/588")
+	require.NoError(b, err)
+	for n := 0; n < b.N; n++ {
+		es.Scale(5)
+	}
+}
+
+func TestMain(m *testing.M) {
+	rand.Seed(time.Now().UnixNano())
```

Comment:
I do prefer if we at least log it somehow so we can at least see it in case it produces an error but just in general it doesn't 
I would also really prefer if we actually don't use `TestMain`, but instead each test has it's own source as I've done in https://github.com/loadimpact/k6/blob/af4321099e8f65f9279e92c8ec46a6f7d3c1cc29/lib/execution_segment_test.go#L493-L494 maybe we can make it use `time.Now()` for tests and keep it using a constant for the benchmarks?

Reply:
> I do prefer if we at least log it somehow so we can at least see it in case it produces an error but just in general it doesn't

You mean log the chosen seed? Sure, I'll add it.


> I would also really prefer if we actually don't use `TestMain`, but instead each test has it's own source as I've done in

What would be the reason for that? In this case neither test "cares" about its random source, it just needs it to use a different seed on each run, so I don't see what we would gain by initializing a new source separately. After all, tests that need a constant seed can initialize their own and not use the global source, like your example.

> maybe we can make it use `time.Now()` for tests and keep it using a constant for the benchmarks?

The benchmark here doesn't depend on `rand`, but yeah, benchmarks that do should use a constant seed.